x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
-			compatibleVerLookup[compatibleVersions[i]] = true
+		if identifier == supportedIdentifier {
+			return supportedVersion, true
+		}
+	}
+	return "", false
+}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
+// PickVersion iterates over the descending ordered set of compatible IBC
+// versions and selects the first version with a version identifier that is
+// supported by the counterparty. The returned version contains a feature
+// set with the intersection of the features supported by the source and
+// counterparty chains. This function is called in the ConnOpenTry handshake
+// procedure.
+func PickVersion(counterpartyVersions []string) (string, error) {
+	versions := GetCompatibleVersions()
+
+	for _, ver := range versions {
+		// check if the source version is supported by the counterparty
+		if counterpartyVer, found := FindSupportedVersion(ver, counterpartyVersions); found {
+			sourceIdentifier, sourceFeatures, err := UnpackVersion(ver)
+			if err != nil {
+				return "", err
 			}
 
-			if _, ok := counterpartyVerLookup[compatibleVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, compatibleVersions[i])
+			_, counterpartyFeatures, err := UnpackVersion(counterpartyVer)
+			if err != nil {
+				return "", err
 			}
+
+			featureSet := GetFeatureSetIntersection(sourceFeatures, counterpartyFeatures)
+
+			version := CreateVersionString(sourceIdentifier, featureSet)
+			return version, nil
 		}
+	}
+
+	return "", sdkerrors.Wrapf(
+		ErrVersionNegotiationFailed,
+		"failed to find a matching counterparty version (%s) from the supported version list (%s)", counterpartyVersions, versions,
+	)
+}
 
-	case n < m:
-		for i := m - 1; i >= n; i-- {
-			compatibleVerLookup[compatibleVersions[i]] = true
+// GetFeatureSetIntersection returns the intersections of source feature set
+// and the counterparty feature set. This is done by iterating over all the
+// features in the source version and seeing if they exist in the feature
+// set for the counterparty version.
+func GetFeatureSetIntersection(sourceFeatureSet, counterpartyFeatureSet []string) (featureSet []string) {
+	for _, feature := range sourceFeatureSet {
+		if contains(feature, counterpartyFeatureSet) {
+			featureSet = append(featureSet, feature)
 		}
+	}
 
-		for i := n - 1; i >= 0; i-- {
-			counterpartyVerLookup[counterpartyVersions[i]] = true
-			compatibleVerLookup[compatibleVersions[i]] = true
+	return featureSet
+}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
-			}
+// VerifyProposedFeatureSet verifies that the entire feature set in the
+// proposed version is supported by this chain.
+func VerifyProposedFeatureSet(proposedVersion, supportedVersion string) bool {
```

Comment:
Looks like this function will pass even if the proposedVersion has an empty feature set. Is this intended behaviour?

Should we allow connections to be established with no intersecting feature set? My thinking is no.

Reply:
I think we should allow zero-intersection, connections alone are useful for other kinds of state verification

Reply:
this is intended behaviour. feature sets are extra on top of the core protocol